### DISCUSSION TITLE
[Offer] Some help developing new features

### DISCUSSION THREAD

**Discussion by fjpolo** [nand2mario/nestang#42] [General]
Hi, I love the idea of NESTang and got myself one a few weeks back and started tinkering with it, currently working on supporting the originall NES gampead: [fjpolo/TangNano20k_NESGamepad/src/NESgamepad](https://github.com/fjpolo/nestang/tree/fjpolo/TangNano20k_NESGamepad/src/NESgamepad)

I had some ideas to work on in the future:
- Cheat codes, supporting a cheat file (hardcoded cheats version: [dev_CheatWizard](https://github.com/fjpolo/nestang/tree/dev_CheatWizard))
- SRAM save support
- Support for illegal opcodes, I believe some ROMs (such as NESFab ROMs) are unsupported because of this
- Possibly savestates, though it might be a lot of work
- Support for FamiCom cartridges (if I find suitable bidirectional serial to parallel converters)

So, if it sounds I could hel out, I'd love to be part of the team and help mantain NESTang :)

**Comment by nand2mario** [comment]
Hi fjpolo,

Help is always welcome!

NES gamepad support sounds good. What circuitry does your NESgamepad need, apart from the connector itself? I have a pair of these pads somewhere. So I'd be interested in trying it. (There's an issue #12 on NES controller support.)

Cheats and save support are also often-requested features. So those sound good too.

**Comment by nand2mario** [comment]
Just saw you comments in #41. I'll take a look at the required circuitry.

**Comment by fjpolo** [comment]
Well I tried a simple 2R/R voltage divider for 5v to 3V3 and this: [this](https://next-hack.com/index.php/2020/02/15/how-to-interface-a-3-3v-output-to-a-5v-input/#:~:text=generic%20%E2%80%9CR%E2%80%9D%20value.-,Fig.%209.,-Adding%20R3%20will) for 3V3 to 5V connection but didn't work, signal was close to a sine that square, probably because of rise and fall time of the current through the Rs. Now waiting for some logic shifters to test again.

Cheats is gonna be fun to play with, it's really early dev, just hardcoded. Thinking about a way to save the ROM name and check if a cheat file is available for that ROM, or something like that, since cheat file can be a binary file with address, data and compare data, pretty easy

**Comment by fjpolo** [comment]
Doesn't really make sense to keep this one alive :)
